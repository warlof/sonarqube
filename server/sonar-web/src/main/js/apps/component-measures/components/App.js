/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import React from 'react';
import Helmet from 'react-helmet';
import MeasureContent from './MeasureContent';
import Sidebar from '../sidebar/Sidebar';
import { parseQuery, serializeQuery } from '../utils';
import { translate } from '../../../helpers/l10n';
import type { Component, Query, Period } from '../types';
import type { RawQuery } from '../../../helpers/query';
import type { Metric } from '../../../store/metrics/actions';
import type { MeasureEnhanced } from '../../../components/measure/types';
import '../style.css';

type Props = {|
  component: Component,
  location: { pathname: string, query: RawQuery },
  fetchMeasures: (
    Component,
    Array<string>
  ) => Promise<{ component: Component, measures: Array<MeasureEnhanced>, leakPeriod: ?Period }>,
  fetchMetrics: () => void,
  metrics: { [string]: Metric },
  metricsKey: Array<string>,
  router: {
    push: ({ pathname: string, query?: RawQuery }) => void
  }
|};

type State = {|
  loading: boolean,
  measures: Array<MeasureEnhanced>,
  leakPeriod: ?Period
|};

export default class App extends React.PureComponent {
  mounted: boolean;
  props: Props;
  state: State;

  constructor(props: Props) {
    super(props);
    this.state = {
      loading: true,
      measures: [],
      leakPeriod: null
    };
  }

  componentDidMount() {
    this.mounted = true;
    this.props.fetchMetrics();
    this.fetchMeasures(this.props);

    const footer = document.getElementById('footer');
    if (footer) {
      footer.classList.add('search-navigator-footer');
    }
  }

  componentWillReceiveProps(nextProps: Props) {
    if (
      nextProps.component.key !== this.props.component.key ||
      nextProps.metrics !== this.props.metrics
    ) {
      this.fetchMeasures(nextProps);
    }
  }

  componentWillUnmount() {
    this.mounted = false;
    const footer = document.getElementById('footer');
    if (footer) {
      footer.classList.remove('search-navigator-footer');
    }
  }

  fetchMeasures = ({ component, fetchMeasures, metrics, metricsKey }: Props) => {
    this.setState({ loading: true });
    const filterdKeys = metricsKey.filter(
      key => !metrics[key].hidden && !['DATA', 'DISTRIB'].includes(metrics[key].type)
    );
    fetchMeasures(component.key, filterdKeys).then(
      ({ measures, leakPeriod }) => {
        if (this.mounted) {
          this.setState({
            loading: false,
            leakPeriod,
            measures: measures.filter(measure => measure.value != null || measure.leak != null)
          });
        }
      },
      () => this.setState({ loading: false })
    );
  };

  updateQuery = (newQuery: Query) => {
    const query = serializeQuery({
      ...parseQuery(this.props.location.query),
      ...newQuery
    });
    this.props.router.push({
      pathname: this.props.location.pathname,
      query: {
        ...query,
        id: this.props.component.key
      }
    });
  };

  render() {
    const isLoading = this.state.loading || this.props.metricsKey.length <= 0;
    if (isLoading) {
      return <i className="spinner spinner-margin" />;
    }
    const query = parseQuery(this.props.location.query);
    const metric = this.props.metrics[query.metric];
    return (
      <div className="layout-page" id="component-measures">
        <Helmet title={translate('layout.measures')} />

        <div className="layout-page-side-outer">
          <div className="layout-page-side" style={{ top: 95 }}>
            <div className="layout-page-side-inner">
              <div className="layout-page-filters">
                <Sidebar
                  measures={this.state.measures}
                  selectedMetric={query.metric}
                  updateQuery={this.updateQuery}
                />
              </div>
            </div>
          </div>
        </div>

        {metric != null &&
          <MeasureContent
            className="layout-page-main-inner"
            rootComponent={this.props.component}
            fetchMeasures={this.props.fetchMeasures}
            leakPeriod={this.state.leakPeriod}
            metric={metric}
            metrics={this.props.metrics}
            selected={query.selected}
            updateQuery={this.updateQuery}
          />}
      </div>
    );
  }
}
