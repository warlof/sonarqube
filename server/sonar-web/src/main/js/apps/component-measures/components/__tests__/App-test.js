/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import { shallow } from 'enzyme';
import App from '../App';

const METRICS = {
  lines_to_cover: {
    key: 'lines_to_cover',
    type: 'INT',
    name: 'Lines to Cover',
    domain: 'Coverage'
  },
  coverage: { key: 'coverage', type: 'PERCENT', name: 'Coverage', domain: 'Coverage' },
  duplicated_lines_density: {
    key: 'duplicated_lines_density',
    type: 'PERCENT',
    name: 'Duplicated Lines (%)',
    domain: 'Duplications'
  },
  new_bugs: { key: 'new_bugs', type: 'INT', name: 'New Bugs', domain: 'Reliability' }
};

const PROPS = {
  component: { key: 'foo' },
  location: { pathname: '/component_measures', query: { metric: 'coverage' } },
  fetchMeasures: () => {},
  fetchMetrics: () => {},
  metrics: METRICS,
  metricsKey: ['lines_to_cover', 'coverage', 'duplicated_lines_density', 'new_bugs'],
  router: { push: () => {} }
};

it('should render correctly', () => {
  const wrapper = shallow(<App {...PROPS} />);
  expect(wrapper.find('.spinner')).toHaveLength(1);
  wrapper.setState({ loading: false });
  expect(wrapper).toMatchSnapshot();
});
