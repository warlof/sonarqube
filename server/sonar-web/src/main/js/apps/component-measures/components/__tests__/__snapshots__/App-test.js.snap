// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="layout-page"
  id="component-measures"
>
  <HelmetWrapper
    encodeSpecialCharacters={true}
    title="layout.measures"
  />
  <div
    className="layout-page-side-outer"
  >
    <div
      className="layout-page-side"
      style={
        Object {
          "top": 95,
        }
      }
    >
      <div
        className="layout-page-side-inner"
      >
        <div
          className="layout-page-filters"
        >
          <Sidebar
            measures={Array []}
            selectedMetric="coverage"
            updateQuery={[Function]}
          />
        </div>
      </div>
    </div>
  </div>
  <MeasureContent
    className="layout-page-main-inner"
    fetchMeasures={[Function]}
    leakPeriod={null}
    metric={
      Object {
        "domain": "Coverage",
        "key": "coverage",
        "name": "Coverage",
        "type": "PERCENT",
      }
    }
    metrics={
      Object {
        "coverage": Object {
          "domain": "Coverage",
          "key": "coverage",
          "name": "Coverage",
          "type": "PERCENT",
        },
        "duplicated_lines_density": Object {
          "domain": "Duplications",
          "key": "duplicated_lines_density",
          "name": "Duplicated Lines (%)",
          "type": "PERCENT",
        },
        "lines_to_cover": Object {
          "domain": "Coverage",
          "key": "lines_to_cover",
          "name": "Lines to Cover",
          "type": "INT",
        },
        "new_bugs": Object {
          "domain": "Reliability",
          "key": "new_bugs",
          "name": "New Bugs",
          "type": "INT",
        },
      }
    }
    rootComponent={
      Object {
        "key": "foo",
      }
    }
    selected=""
    updateQuery={[Function]}
  />
</div>
`;
